utils/staking-miner/src/monitor.rs

```diff
@@ -131,39 +209,58 @@ macro_rules! monitor_cmd_for { ($runtime:tt) => { paste::paste! {
 			config: MonitorConfig,
 		) {
 
+			async fn flatten<T>(
+				handle: tokio::task::JoinHandle<Result<T, StakingMinerError>>
+			) -> Result<T, StakingMinerError> {
+				match handle.await {
+					Ok(Ok(result)) => Ok(result),
+					Ok(Err(err)) => Err(err),
+					Err(err) => panic!("tokio spawn task failed; kill task: {:?}", err),
+				}
+			}
+
 			let hash = at.hash();
 			log::trace!(target: LOG_TARGET, "new event at #{:?} ({:?})", at.number, hash);
 
-			// if the runtime version has changed, terminate.
+			// block on this because if this fails there is no way to recover from
+			// that error i.e, upgrade/downgrade required.
 			if let Err(err) = crate::check_versions::<Runtime>(&rpc).await {
```

Comment:
imo, if possible we should try and put this in the try join and then have a special match case for the `check_versions` error where we do `tx.send(err.into()); return;`

Reply:
the issue is that  `try_join` will return as soon one future fails i.e, returns `Err()`

so this might fail but something else always get first and the main task will never get notified that 
the runtime version is outdated and it will try to spawn new tasks that always will fail..

that's the reason why it did it like this but I guess I could do `join(check_version, try_join(a, b, c))` instead

Reply:
I realized it would be slightly better to have a runtime version subscription here instead of polling `runtime_version` all over the place, when it received a new version -> quit all

Reply:
unfortunately I ran into https://github.com/paritytech/substrate/issues/10910 occasionally when I tested it